Code Clean Up (#2)
brett060102/ansible-container: README.md

# Containerized Ansible: What's inside #

This container provides the ansible toolstack inside a container.
* '''Dockerfile''' with the definition of the ansible container
  * based on OpenSUSE Tumbleweed
  * installs ansible and some additional tools
## System Setup ##
* Make sure needed packages installed on the host

>   * sudo transactional-update pkg install podman python310-rpm

Reviewer: podman is part of the ALP Base - no need to install it

We probably should have this as a list of packages, with corresponding annotations as to what in Ansible depends on the relevant package, e.g. the `package_facts` module depends on this being installed on RPM package bases systems...

Author: I know, but since it is required I wanted  to have it in case someone else used this as basis on a different system.
Not sure I understand the last part python310-rpm is needed so ansible can querry what is installed.

Reviewer: For target system, which is a container host OS that comes with podman already installed, explicitly saying to install podman is a redundant operation. However if someone wanted to use this container on a different platform, yes they would need to install podman, but the command to do so likely wouldn't be using transactional-update.

  * system reboot is required after all transactional updates
    * sudo shutdown -r now

## Ansible commands
The ansible commands are provided as symlinks to ansible-wrapper.sh.

> The commands will instanciate the container and execute the ansible

Reviewer: ```suggestion
The commands will instantiate the container and execute the ansible
```

command.

## To install ansible commands ##

* as root:
  * for the root user the ansible commands are placed in /usr/local/bin

>   * (cd /; podman container runlabel install ansible)

Reviewer: The system install should not require being cd'd to `/`... If it does we should fix that...

Reviewer: ```suggestion
  * podman container runlabel install ansible
```

* as non-root

>   * for non-root users the ansible commands are placed in the user's ~/bin

Reviewer: Since we require the user to cd to ~ before running the podman command, this line probably should look something like:

```suggestion
  * for non-root users the ansible commands should be placed in the user's ~/bin, so changes to the users home directory first.
```

Author: "should placed" is not quite right. If the example is followed, then the commands will be placed in the user ~/bin area. 

I could expand the section to read:
For non-root users  'podman container runlabel user-install ansible' will place the ansible commands in ${PWD}/bin. The following will install the ansible commands into the current users bin area (~/bin).

  * (cd ~; podman container runlabel user-install ansible)

## Ansible Commands ##
* ansible
* ansible-community
* ansible-config
* ansible-connection
* ansible-console
* ansible-doc
* ansible-galaxy
* ansible-inventory
* ansible-playbook
* ansible-pull

> * ansible-vault

Reviewer: Missing `ansible-test` here...

```suggestion
* ansible-test
* ansible-vault
```

 
## Uninstall ansible commands  ##
* as root:

>   * (cd /; podman container runlabel uninstall ansible)

Reviewer: ```suggestion
  * podman container runlabel uninstall ansible
```

* as non-root
  * (cd ~; podman container runlabel user-uninstall ansible)

## Operation is through SSH back to container host or to other remote systems  ##
Since ansible is running within a container, the local host is the container and not
system instantiating the container. Any changes made to the local host are made to the
container and would be lost when the container exits. To make host running the container
host.containers.internal can be used in inventory record like:
```
inventory.yaml:

> all:

Reviewer: We probably should change this here, and in the example, to look something more like:

```
alphost_group:
  hosts:
    alphost:
      ansible_host: host.containers.internal
```

I'll put up a PR to address that...

The reason for doing so is that technically `all` is a special group that is owned by the ansible inventory management system; any host that is defined in inventory gets automatically added to the `all` group.

We potentially could also define this in the /etc/ansible/hosts file in the container image by appending the following lines:

```
# Define the ALP podman host
[alphost_group]
alphost ansible_host=host.containers.internal
```

And then we just use `alphost` as a play target anywhere we need to run against the host.

Reviewer: Have proposed #4 to address this.

  hosts:
      host.containers.internal:
```

> The inventory record could also contaon other hosts to be managed.

Reviewer: ```suggestion
The inventory record could also contain other hosts to be managed.
```

### SSH keys must be set up ###
The container must be able to SSH to the system being managed. So, the system must support SSH access and

> the SSH keys must have been created and the public key must be in the .ssh/authorized_keys file for the

Reviewer: ```suggestion
the SSH keys must have been created (using `ssh-keygen`) and the public key must be in the `.ssh/authorized_keys` file for the
```

> target user. While the root user can be used so long as teh system allows SSHing to the root account,

Reviewer: ```suggestion
target user. While the root user can be used so long as the system allows SSH'ing to the root account,
```

the preferred method to to use an non-root account that has passwordless sudo rights. Anny operation in ansible

> play books that require system privilege would then to use "become: true"

Reviewer: ```suggestion
play books that require system privilege would then need to use "become: true"
```

> SSH access can ve validated with "$ ssh localhost"

Reviewer: ```suggestion
SSH access can be validated with `ssh localhost`.
```
